sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala

```diff
@@ -17,53 +17,234 @@
 
 package org.apache.spark.sql.catalyst.types
 
-import org.apache.spark.sql.types._
-
+import scala.reflect.runtime.universe.TypeTag
+import scala.reflect.runtime.universe.typeTag
+import scala.util.control.NonFatal
+import org.apache.spark.sql.catalyst.util.SQLOrderingUtil
+import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.sql.types.{AtomicType, BinaryType, BooleanType, ByteType, DataType, DateType, DayTimeIntervalType, Decimal, DecimalType, DoubleType, FloatType, IntegerType, LongType, NullType, NumericType, ShortType, StringType, StructField, TimestampNTZType, TimestampType}
+import org.apache.spark.unsafe.types.{ByteArray, UTF8String}
 
 sealed abstract class PhysicalDataType
 
-sealed abstract class PhysicalPrimitiveType extends PhysicalDataType
+object PhysicalDataType {
+  def apply(dt: DataType): PhysicalDataType = dt match {
+    case NullType => PhysicalNullType
+    case BooleanType => PhysicalBooleanType
+    case ByteType => PhysicalByteType
+    case ShortType => PhysicalShortType
+    case IntegerType => PhysicalIntegerType
+    case LongType => PhysicalLongType
+    case FloatType => PhysicalFloatType
+    case DoubleType => PhysicalDoubleType
+    case DecimalType.Fixed(p, s) => PhysicalDecimalType(p, s)
+    case BinaryType => PhysicalBinaryType
+    case _ => UninitializedPhysicalType
+  }
+}
+
+trait PhysicalPrimitiveType
+
+sealed abstract class TypedPhysicalDataType extends PhysicalDataType {
+  private[sql] type InternalType
+}
+
+sealed abstract class OrderedPhysicalDataType extends TypedPhysicalDataType {
+  private[sql] def ordering: Ordering[InternalType]
+}
+
+object OrderedPhysicalDataType {
+  def apply(dt: DataType): OrderedPhysicalDataType =
+    PhysicalDataType(dt).asInstanceOf[OrderedPhysicalDataType]
+
+  def ordering(dt: DataType): Ordering[Any] = {
+    try apply(dt).ordering.asInstanceOf[Ordering[Any]] catch {
+      case NonFatal(_) =>
+        throw QueryExecutionErrors.unsupportedTypeError(dt)
+    }
+  }
+}
+
+sealed abstract class PhysicalAtomicType extends OrderedPhysicalDataType {
+  private[sql] val tag: TypeTag[InternalType]
+}
+
+object PhysicalAtomicType extends PhysicalAtomicType {
```

Comment:
Why is this extending `PhysicalAtomicType`? This is an abstract data type if any, not a concrete one.

Reply:
No you don't need that. `PhysicalAtomicType.apply(dt)` returns an instance of `PhysicalAtomicType`, and not the companion; just remove it.

Reply:
I just use `OrderedPhysicalDataType` then